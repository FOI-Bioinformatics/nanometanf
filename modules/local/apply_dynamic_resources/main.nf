process APPLY_DYNAMIC_RESOURCES {
    tag "$meta.id"
    label 'process_single'

    input:
    tuple val(meta), path(allocation_file)
    val target_process_name

    output:
    tuple val(meta), val(dynamic_config), emit: config
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    """
    #!/usr/bin/env python3
    
    import json
    import os
    from pathlib import Path
    
    meta = ${groovy.json.JsonBuilder(meta).toString()}
    target_process = "${target_process_name}"
    
    # Load optimal allocation
    with open('${allocation_file}', 'r') as f:
        allocation = json.load(f)
    
    # Extract the Nextflow process configuration
    dynamic_config = allocation.get('nextflow_process_config', {})
    
    print(f"Dynamic resource configuration for {meta['id']} -> {target_process}:")
    for key, value in dynamic_config.items():
        print(f"  {key}: {value}")
    
    # Generate versions file
    with open('versions.yml', 'w') as f:
        f.write('''\"${task.process}\":
    python: \"3.9\"
    dynamic_resources: \"1.0\"''')
    """

    stub:
    """
    echo '{}' > stub_config.json
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: "3.9"
        dynamic_resources: "1.0"
    END_VERSIONS
    """
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    HELPER FUNCTIONS FOR DYNAMIC RESOURCE APPLICATION
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

def applyDynamicResources(process_config, sample_meta = null, default_config = [:]) {
    /*
    Apply dynamic resource configuration to a process
    This function merges dynamic configs with default configs and applies them
    */
    
    // Start with default configuration
    def final_config = [:]
    final_config.putAll(default_config)
    
    // Apply dynamic configuration if available
    if (process_config && process_config.size() > 0) {
        final_config.putAll(process_config)
        log.info "Applied dynamic resource allocation for ${sample_meta?.id ?: 'unknown'}: ${final_config}"
    } else {
        log.debug "Using default resource configuration: ${final_config}"
    }
    
    return final_config
}

def configureDynamicProcess(process_name, resource_configs, sample_meta = null) {
    /*
    Configure a specific process with dynamic resources
    */
    
    def process_configs = [:]
    
    resource_configs.subscribe { meta, config ->
        if (meta.id == sample_meta?.id) {
            process_configs[process_name] = config
            
            // Apply the configuration to the process
            withName:process_name {
                cpus = config.cpus ?: 1
                memory = config.memory ?: '4.GB'
                time = config.time ?: '6.h'
                
                if (config.accelerator) {
                    accelerator = config.accelerator
                }
                
                if (config.errorStrategy) {
                    errorStrategy = config.errorStrategy
                }
                
                if (config.maxRetries) {
                    maxRetries = config.maxRetries
                }
            }
        }
    }
    
    return process_configs
}

def createResourceSelector(ch_resource_configs) {
    /*
    Create a resource selector that processes can use to get their optimal configuration
    */
    
    def resource_selector = ch_resource_configs
        .map { meta, config_file ->
            // Parse the configuration file to extract the nextflow config
            def config = readJSON(file: config_file)
            def nextflow_config = config.nextflow_process_config ?: [:]
            
            [ meta, nextflow_config ]
        }
    
    return resource_selector
}

def getOptimalResourceConfig(sample_meta, tool_name, ch_resource_configs, default_config = [:]) {
    /*
    Get optimal resource configuration for a specific sample and tool
    Returns a channel with the configuration
    */
    
    def optimal_config = ch_resource_configs
        .filter { meta, config -> meta.id == sample_meta.id }
        .map { meta, config ->
            // Apply tool-specific adjustments if needed
            def adjusted_config = config.clone()
            
            // Tool-specific resource adjustments
            switch(tool_name) {
                case 'KRAKEN2':
                    // Ensure minimum memory for Kraken2
                    def memory_val = config.memory?.replaceAll(/[^0-9.]/, '')?.toFloat() ?: 4
                    if (memory_val < 16) {
                        adjusted_config.memory = '16.GB'
                    }
                    break
                    
                case 'DORADO_BASECALLER':
                    // Ensure GPU configuration is properly set
                    if (config.accelerator) {
                        adjusted_config.clusterOptions = '--gres=gpu:1'
                    }
                    break
                    
                case 'FLYE':
                case 'MINIASM':
                    // Assembly tools need more time
                    def time_val = config.time?.replaceAll(/[^0-9.]/, '')?.toFloat() ?: 6
                    adjusted_config.time = "${Math.max(time_val, 12)}.h"
                    break
            }
            
            // Merge with defaults
            def final_config = [:]
            final_config.putAll(default_config)
            final_config.putAll(adjusted_config)
            
            [ meta, final_config ]
        }
        .ifEmpty { [ sample_meta, default_config ] }
    
    return optimal_config
}