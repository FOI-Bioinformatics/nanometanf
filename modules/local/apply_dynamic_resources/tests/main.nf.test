nextflow_process {

    name "Test APPLY_DYNAMIC_RESOURCES"
    script "../main.nf"
    process "APPLY_DYNAMIC_RESOURCES"

    tag "modules"
    tag "modules_local"
    tag "apply_dynamic_resources"

    test("Should apply basic resource adjustments") {

        tag "basic_adjustment"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'basic_adjustment_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/allocations/balanced_allocation.json')
                ]
                input[1] = [
                    safety_factor: 0.8,
                    apply_process_specific: true,
                    respect_system_limits: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle aggressive resource scaling") {

        tag "aggressive_scaling"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'aggressive_scaling_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/allocations/balanced_allocation.json')
                ]
                input[1] = [
                    safety_factor: 0.9,
                    apply_process_specific: true,
                    respect_system_limits: true,
                    allow_aggressive_scaling: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should respect system resource limits") {

        tag "system_limits"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'constrained_system_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/allocations/balanced_allocation.json')
                ]
                input[1] = [
                    safety_factor: 0.8,
                    apply_process_specific: true,
                    respect_system_limits: true,
                    enforce_hard_limits: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle GPU resource allocation") {

        tag "gpu_allocation"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'gpu_allocation_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/allocations/balanced_allocation.json')
                ]
                input[1] = [
                    safety_factor: 0.8,
                    apply_gpu_optimization: true,
                    gpu_allocation_strategy: "balanced"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should generate conservative resource allocation") {

        tag "conservative_allocation"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'conservative_allocation_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/allocations/balanced_allocation.json')
                ]
                input[1] = [
                    safety_factor: 0.6,
                    apply_process_specific: true,
                    conservative_mode: true,
                    respect_shared_resources: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle invalid or malformed recommendations") {

        tag "edge_cases"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'invalid_input_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/edge_cases/extreme_allocation.json')
                ]
                input[1] = [
                    safety_factor: 0.8,
                    handle_invalid_input: true,
                    use_fallback_defaults: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

}