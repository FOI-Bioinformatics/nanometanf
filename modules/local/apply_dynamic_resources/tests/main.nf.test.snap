{
    "Should handle aggressive resource scaling": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "aggressive_scaling_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ],
                "config": [
                    [
                        {
                            "id": "aggressive_scaling_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T07:44:55.172761"
    },
    "Should handle GPU resource allocation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "gpu_allocation_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ],
                "config": [
                    [
                        {
                            "id": "gpu_allocation_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T07:45:01.455982"
    },
    "Should generate conservative resource allocation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "conservative_allocation_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ],
                "config": [
                    [
                        {
                            "id": "conservative_allocation_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T07:45:04.604757"
    },
    "Should apply basic resource adjustments": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "basic_adjustment_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ],
                "config": [
                    [
                        {
                            "id": "basic_adjustment_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T07:44:52.017208"
    },
    "Should respect system resource limits": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "constrained_system_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ],
                "config": [
                    [
                        {
                            "id": "constrained_system_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,048cb9a460187ab65b6a5d390d0b627f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T07:44:58.309075"
    },
    "Should handle invalid or malformed recommendations": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "invalid_input_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,a7ed9bbb2ad46e5c210562c09ca6a00c"
                    ]
                ],
                "1": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ],
                "config": [
                    [
                        {
                            "id": "invalid_input_test",
                            "single_end": true
                        },
                        "dynamic_config.json:md5,a7ed9bbb2ad46e5c210562c09ca6a00c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,40626a5b37af7aec3d03fbe8b3f1ad09"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T07:45:07.755915"
    }
}